check-disk/lib/check-disk.go

```diff
@@ -0,0 +1,212 @@
+		}
+	} else {
+		v, err := strconv.ParseFloat(threshold, 64)
+		if err != nil {
+			return checkers.UNKNOWN, err
+		}
+
+		if v > float64(disk.Free) {
+			current = status
+		}
+	}
+
+	return current, nil
+}
+
+func genMessage(disk *gpud.UsageStat, u unit) string {
+	all := float64(disk.Total) / u.Size
+	used := float64(disk.Used) / u.Size
+	free := float64(disk.Free) / u.Size
+	freePct := float64(100) - disk.UsedPercent
+	inodesFreePct := float64(100) - disk.InodesUsedPercent
+
+	return fmt.Sprintf("Path: %v, All: %.2f %v, Used: %.2f %v, Free: %.2f %v, Free percentage: %.2f (inodes: %.2f)", disk.Path, all, u.Name, used, u.Name, free, u.Name, freePct, inodesFreePct)
+}
+
+// Do the plugin
+func Do() {
+	ckr := run(os.Args[1:])
+	ckr.Name = "Disk"
+	ckr.Exit()
+}
+
+func run(args []string) *checkers.Checker {
+	_, err := flags.ParseArgs(&opts, args)
+	if err != nil {
+		os.Exit(1)
+	}
+
+	partitions, err := listPartitions()
+	if err != nil {
+		return checkers.Unknown(fmt.Sprintf("Faild to fetch partitions: %s", err))
+	}
+
+	if opts.Path != nil {
+		exist := false
+		for _, partition := range partitions {
+			if *opts.Path == partition.Mountpoint {
+				partitions = []gpud.PartitionStat{partition}
+				exist = true
+				break
+			}
+		}
+
+		if !exist {
+			return checkers.Unknown(fmt.Sprintf("Faild to fetch mountpoint: %s", errors.New("Invalid argument flag '-p, --path'")))
+		}
+	}
+
+	if opts.Path == nil && opts.Exclude != nil {
+		var tmp []gpud.PartitionStat
+		for _, partition := range partitions {
+			if *opts.Exclude != partition.Mountpoint {
+				tmp = append(tmp, partition)
+			}
+		}
+		partitions = tmp
+	}
+
+	var disks []*gpud.UsageStat
+
+	for _, partition := range partitions {
+		disk, err := gpud.Usage(partition.Mountpoint)
+		if err != nil {
+			return checkers.Unknown(fmt.Sprintf("Faild to fetch disk usage: %s", err))
+		}
+
+		disks = append(disks, disk)
+	}
+
+	u := unit{"MB", mb}
+	if opts.Units != nil {
+		us := strings.ToLower(*opts.Units)
+		if us == "bytes" {
+			u = unit{us, b}
+		} else if us == "kb" {
+			u = unit{us, mb}
+		} else if us == "gb" {
+			u = unit{us, gb}
+		} else if us == "tb" {
+			u = unit{us, tb}
+		} else {
+			return checkers.Unknown(fmt.Sprintf("Faild to check disk status: %s", errors.New("Invalid argument flag '-u, --units'")))
+		}
+	}
+
+	checkSt := checkers.OK
+	if opts.Warning != nil {
+		for _, disk := range disks {
+			checkSt, err = checkStatus(checkSt, *opts.Warning, u.Size, disk, checkers.WARNING)
+			if err != nil {
+				return checkers.Unknown(fmt.Sprintf("Faild to check disk status: %s", err))
+			}
+
+			if checkSt == checkers.WARNING {
+				break
+			}
+		}
+	}
+
+	if opts.Critical != nil {
+		for _, disk := range disks {
+			checkSt, err = checkStatus(checkSt, *opts.Critical, u.Size, disk, checkers.CRITICAL)
+			if err != nil {
+				return checkers.Unknown(fmt.Sprintf("Faild to check disk status: %s", err))
+			}
+
+			if checkSt == checkers.CRITICAL {
+				break
```

Comment:
[nit]
If you can check critical first,  you can skip checking warnings when critical error happens